Fabulous.XamarinForms/src/Fabulous.XamarinForms.Core/ViewUpdaters.fs

```diff
@@ -28,6 +28,120 @@ module ViewUpdaters =
             if target <> null then currValue.UpdateIncremental(prevValue, target)            
         | ValueSome _, ValueNone ->
             clearValue ()
+           
+    /// Incremental list maintenance: given a collection, and a previous version of that collection, perform
+    /// a reduced number of clear/move/create/update/remove operations
+    ///
+    /// The algorithm will try in priority to update elements sharing the same instance (usually achieved with dependsOn)
+    /// or sharing the same key. All other elements will try to reuse existing previous elements where possible.
+    /// If no reuse is possible, the element will create a new control.
+    let updateCollectionGenericInternal
+            (prevCollOpt: 'T[] voption)
+            (collOpt: 'T[] voption)
+            (getKey: 'T -> string voption)
+            (canReuse: 'T -> 'T -> bool)
+            (move: int -> 'T -> unit)
+            (create: int -> 'T -> unit)
+            (update: int -> 'T -> 'T -> unit)
+            (remove: int -> unit)
+            (clear: unit -> unit) =
+        
+        match prevCollOpt, collOpt with 
+        | ValueSome prevColl, ValueSome newColl when identical prevColl newColl -> ()
+        | ValueSome prevColl, ValueSome newColl when prevColl <> null && newColl <> null && prevColl.Length = 0 && newColl.Length = 0 -> ()
+        | _, ValueNone -> clear ()
+        | _, ValueSome coll when (coll = null || coll.Length = 0) -> clear ()
+        | _, ValueSome coll ->
+            
+            let prevColl =
+                match prevCollOpt with
+                | ValueSome x -> x
+                | _ -> [||]
+
+            // Count the existing targetColl
+            let n = prevColl.Length
+            
+            let availableKeyedElements = Dictionary<string,'T>()
+            let rest= ResizeArray<'T>()
+            let newKeys = HashSet<string>()
+            
+            for i in 0 .. coll.Length-1 do
+                 let newChild = coll.[i]
+                 let key = getKey newChild
+                 match key with
+                 | ValueSome key -> newKeys.Add key |> ignore
+                 | ValueNone _ -> ()
+                 
+            for i in 0..n-1 do
+                let key = getKey prevColl.[i]
+                match key with
+                | ValueSome key ->
+                    availableKeyedElements.[key] <- prevColl.[i]
```

Comment:
I would check if the key is also in `coll` here.
If it's there then add to `availableKeyedElements` otherwise add to `rest`.
It allows you to not recheck it later.

Reply:
Ok